File: story-driven-delivery/skills/abd-acceptance-test-driven-development/templates/acceptance-tests.java
/**
 * ## Instructions (for skill maintainers — do NOT paste this block into project test files)
 *
 * This template defines the required structure for acceptance test files produced by the
 * abd-acceptance-test-driven-development skill using JUnit 5 (Jupiter).
 * When writing test files for a project, omit this Instructions block and emit only the
 * sections below.
 *
 * ### Inputs
 * Work from whatever behavioral context is available:
 * - Spec-by-example scenarios (ideal) — encode Given/When/Then steps directly.
 * - Acceptance criteria (WHEN/THEN) — convert THEN to assertions, derive Given from preconditions.
 * - Stories + notes — identify normal, edge, and failure behaviors; make them concrete in tests.
 * - Unstructured context — extract observable behaviors, name them, then encode as tests.
 *
 * ### Layout
 * 1. Package declaration    — matches project source tree.
 * 2. Imports                — JUnit 5, then third-party, then local; all at top.
 * 3. @DisplayName           — human-readable label on the class (behavior cluster name from source).
 * 4. HELPER METHODS         — private static methods for setup/action/assertion (each under 20 lines).
 * 5. @BeforeEach / @AfterEach — shared setup / teardown for the class.
 * 6. @Test METHODS          — one per scenario or condition.
 *    @DisplayName on each @Test carries the scenario outcome in plain English.
 *    Every method uses // Given / // When / // Then comments and calls helper methods.
 *
 * ### Naming
 * | source level        | test artifact     | example                                    |
 * |---------------------|-------------------|--------------------------------------------|
 * | Area / flow / group | File name         | {AreaPascalCase}Test.java                  |
 * | Behavior / story    | @DisplayName(class)| "Place Order"                              |
 * | Scenario / condition| @DisplayName(method)| "order is confirmed for available items"  |
 * | Scenario / condition| Method name        | orderIsConfirmedForAvailableItems()        |
 */

package {com.example.domain.area};

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
// import {com.example.domain.DomainEntity};

// =============================================================================
// {Area Name} Tests
//
// Behaviors covered:
// - {Story / Behavior Name 1}
// - {Story / Behavior Name 2}
// =============================================================================

@DisplayName("{Area Name}")
class {AreaName}Test {

    // =========================================================================
    // HELPER METHODS -- reusable operations (each under 20 lines)
    // =========================================================================

    private static {DomainEntity} create{DomainEntity}(String name) {
        // Helper: Create domain entity with given name
        return new {DomainEntity}(name);
    }

    private static void verify{DomainState}({DomainEntity} entity, {ExpectedType} expected) {
        // Helper: Verify entity is in expected state
        assertEquals(expected, entity.get{ObservableProperty}());
    }

    // =========================================================================
    // STORY: {Story Name 1}
    // =========================================================================

    @Nested
    @DisplayName("{Story Name 1}")
    class {StoryName1}Tests {

        private {DomainEntity} {entity};

        @BeforeEach
        void setUp() {
            {entity} = create{DomainEntity}("{name}");
        }

        @AfterEach
        void tearDown() {
            // cleanup if needed
        }

        @Test
        @DisplayName("{outcome of scenario in plain English}")
        void {scenarioNameCamelCase}() {
            // SCENARIO: {Scenario Name}
            // GIVEN: {precondition}
            // WHEN: {action}
            // THEN: {observable outcome}

            // Given
            {DomainEntity} {setup} = create{DomainEntity}("{arg}");

            // When
            {ResultType} result = {setup}.{domainOperation}();

            // Then
            verify{DomainState}({setup}, {expectedValue});
        }

        @Test
        @DisplayName("throws {ExceptionName} when {edge condition}")
        void throws{ExceptionName}When{EdgeCondition}() {
            // SCENARIO: {Edge Case Scenario Name}
            // GIVEN: {edge condition}
            // WHEN: {action}
            // THEN: {failure or edge outcome}

            // Given
            {DomainEntity} {entity} = create{DomainEntity}With{EdgeCondition}();

            // When / Then
            assertThrows(
                {ExpectedException}.class,
                () -> {entity}.{domainOperation}()
            );
        }
    }

    // =========================================================================
    // STORY: {Story Name 2}
    // =========================================================================

    @Nested
    @DisplayName("{Story Name 2}")
    class {StoryName2}Tests {

        @Test
        @DisplayName("{outcome of scenario in plain English}")
        void {scenarioNameCamelCase}() {
            // SCENARIO: {Scenario Name}
            // GIVEN: {precondition}
            // WHEN: {action}
            // THEN: {observable outcome}

            // Given
            {DomainEntity} {entity} = create{DomainEntity}("{arg}");

            // When
            {ResultType} result = {entity}.{domainOperation}();

            // Then
            assertEquals({expectedValue}, result);
        }
    }
}

File: story-driven-delivery/skills/abd-acceptance-test-driven-development/templates/acceptance-tests.py
"""
## Instructions (for skill maintainers — do NOT paste this block into project test files)

This template defines the required structure for acceptance test files produced by the
abd-acceptance-test-driven-development skill.  When the skill writes test files for a
project, omit this Instructions section and emit only the sections below.

### Inputs
Work from whatever behavioral context is available:
- Spec-by-example scenarios (ideal) — encode Given/When/Then steps directly.
- Acceptance criteria (WHEN/THEN) — convert THEN to assertions, derive Given from preconditions.
- Stories + notes — identify normal, edge, and failure behaviors; make them concrete in tests.
- Unstructured context — extract observable behaviors, name them, then encode as tests.

### Layout
1. Module docstring  — one-line summary + list of behaviors covered in this file.
2. Imports           — stdlib → third-party → local; all at top.
3. HELPER FUNCTIONS  — reusable setup/action/assertion functions (each under 20 lines).
4. FIXTURES          — pytest fixtures defined in this test file (not conftest).
5. TEST CLASSES      — one class per behavior cluster (story, flow, feature area).
   Named TestExactBehaviorName (PascalCase of the behavior name from source).
   Each class contains one test method per scenario or condition.
   Every test method uses Given/When/Then comments and calls helper functions.

### Naming
| source level        | test artifact | example                                   |
|---------------------|---------------|-------------------------------------------|
| Area / flow / group | File name     | test_<area_snake_case>.py                 |
| Behavior / story    | Class name    | Test<ExactBehaviorName>                   |
| Scenario / condition| Method name   | test_<outcome_snake_case>                 |
"""

# =============================================================================
# {Sub-Epic Name} Tests
#
# Tests for all stories in '{Sub-Epic Name}' sub-epic:
# - {Story Name 1}
# - {Story Name 2}
# =============================================================================

import pytest
from pathlib import Path
# from <project>.module import ClassUnderTest

# =============================================================================
# HELPER FUNCTIONS — reusable operations (each under 20 lines)
# =============================================================================

def create_{domain_entity}({param}: {type}, workspace: Path) -> {ReturnType}:
    """Helper: Create {domain entity} with {description}."""
    # ... setup ...
    return {domain_entity}


def verify_{domain_state}({entity}: {Type}, {expected}: {type}) -> None:
    """Helper: Verify {domain entity} is in {expected state}."""
    assert {entity}.{observable_property} == {expected}


# =============================================================================
# FIXTURES — defined here, not in conftest
# =============================================================================

@pytest.fixture
def workspace_root(tmp_path: Path) -> Path:
    """Fixture: Temporary workspace directory."""
    workspace = tmp_path / "workspace"
    workspace.mkdir()
    return workspace


# =============================================================================
# STORY: {Story Name 1}
# =============================================================================

class Test{StoryName1}:
    """{Story Name 1} — {one-line description}."""

    def test_{scenario_name}(self, workspace_root: Path) -> None:
        """
        SCENARIO: {Scenario Name}
        GIVEN: {precondition}
        WHEN: {action}
        THEN: {observable outcome}
        """
        # Given
        {entity} = create_{domain_entity}({arg}, workspace_root)

        # When
        result = {entity}.{domain_operation}()

        # Then
        verify_{domain_state}({entity}, {expected_value})

    def test_{edge_case_scenario}(self, workspace_root: Path) -> None:
        """
        SCENARIO: {Edge Case Scenario Name}
        GIVEN: {edge condition}
        WHEN: {action}
        THEN: {failure or edge outcome}
        """
        # Given
        {entity} = create_{domain_entity}_with_{edge_condition}(workspace_root)

        # When / Then
        with pytest.raises({ExpectedException}):
            {entity}.{domain_operation}()


# =============================================================================
# STORY: {Story Name 2}
# =============================================================================

class Test{StoryName2}:
    """{Story Name 2} — {one-line description}."""

    def test_{scenario_name}(self, workspace_root: Path) -> None:
        """
        SCENARIO: {Scenario Name}
        GIVEN: {precondition}
        WHEN: {action}
        THEN: {observable outcome}
        """
        # Given
        {entity} = create_{domain_entity}({arg}, workspace_root)

        # When
        result = {entity}.{domain_operation}()

        # Then
        assert result == {expected_value}
